rafthttp/peer.go

```diff
@@ -15,158 +15,196 @@
+type peer struct {
+	to types.ID
 	r  Raft
-	fs *stats.FollowerStats
 
-	batcher     *Batcher
-	propBatcher *ProposalBatcher
-
-	pipeline *pipeline
-	stream   *stream
-
-	paused  bool
-	stopped bool
+	msgAppDialer  *streamDialer
+	msgAppReader  *streamReader
+	msgAppDecoder *msgAppDecoder
+	msgAppWriter  *streamWriter
+	dialer        *streamDialer
+	reader        *streamReader
+	writer        *streamWriter
+	pipeline      *pipeline
+
+	sendc   chan raftpb.Message
+	recvc   chan raftpb.Message
+	updatec chan string
+	pausec  chan struct{}
+	resumec chan struct{}
+	stopc   chan struct{}
+	done    chan struct{}
 }
 
-func NewPeer(tr http.RoundTripper, u string, id types.ID, cid types.ID, r Raft, fs *stats.FollowerStats, errorc chan error) *peer {
-	return &peer{
-		id:          id,
-		cid:         cid,
-		tr:          tr,
-		u:           u,
-		r:           r,
-		fs:          fs,
-		pipeline:    newPipeline(tr, u, id, cid, fs, errorc),
-		stream:      &stream{},
-		batcher:     NewBatcher(100, appRespBatchMs*time.Millisecond),
-		propBatcher: NewProposalBatcher(100, propBatchMs*time.Millisecond),
+func startPeer(tr http.RoundTripper, u string, local, to, cid types.ID, r Raft, fs *stats.FollowerStats, errorc chan error) *peer {
+	p := &peer{
+		to:            to,
+		r:             r,
+		msgAppDialer:  newStreamDialer(u, tr, local, to, cid, streamTypeMsgApp),
+		msgAppDecoder: newMsgAppDecoder(local, to),
+		msgAppWriter:  startStreamWriter(newMsgAppEncoder(to, fs)),
+		dialer:        newStreamDialer(u, tr, local, to, cid, streamTypeGeneral),
+		writer:        startStreamWriter(&generalEncoder{}),
+		pipeline:      newPipeline(tr, u, to, cid, fs, errorc),
+
+		sendc:   make(chan raftpb.Message),
+		recvc:   make(chan raftpb.Message, recvBufSize),
+		updatec: make(chan string),
+		pausec:  make(chan struct{}),
+		resumec: make(chan struct{}),
+		stopc:   make(chan struct{}),
+		done:    make(chan struct{}),
 	}
+	p.msgAppReader = startStreamReader(p.msgAppDialer, p.msgAppDecoder, p.recv)
+	p.reader = startStreamReader(p.dialer, &generalDecoder{}, p.recv)
+	go p.run()
+	return p
 }
 
-func (p *peer) Update(u string) {
-	p.Lock()
-	defer p.Unlock()
-	if p.stopped {
-		// TODO: not panic here?
-		panic("peer: update a stopped peer")
+func (p *peer) run() {
+	var paused bool
+	// non-blocking run loop
+	for {
+		select {
+		case m := <-p.sendc:
+			if paused {
+				continue
+			}
+			var writec chan raftpb.Message
+			var name string
+			if p.msgAppWriter.canSend(m) {
+				writec = p.msgAppWriter.msgc
+				name = "msgapp stream"
+			} else if p.writer.canSend(m) {
+				writec = p.writer.msgc
+				name = "general stream"
+			} else {
+				writec = p.pipeline.msgc
+				name = "pipeline"
+			}
+			select {
+			case writec <- m:
+			default:
+				log.Printf("peer: dropping %s because maximal number %d of %s buffer entries to %s has been reached",
+					m.Type, streamBufSize, name, p.to)
+			}
+		case mm := <-p.recvc:
+			p.msgAppDecoder.updateTerm(mm.Term)
```

Comment:
i don't quite follow.
maybe only letting msgapp or msgheartbeat update it is better. These messages indicate that the leader is communicating with me.